### PR TITLE
feat: Detect CDK project and add warning message for unsupported commands

### PR DESCRIPTION
**Pull request by mildaniel**
#### Checklist

- [ ] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [ ] Write unit tests
- [ ] Write/update functional tests
- [ ] Write/update integration tests
- [ ] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

**Review by wchengru — APPROVED** [review]

File: samcli/commands/_utils/cdk_support_decorators.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+            ctx = Context.get_current_context()
+
+            try:
+                template_dict = ctx.template_dict
+            except AttributeError:
+                LOG.debug("Ignoring CDK project check as template is not provided in context.")
+                return func(*args, **kwargs)
+
+            if is_cdk_project(template_dict):
+                click.secho("Warning: CDK apps are not officially supported with this command.", fg="yellow")
```
[the remark this answers is not in the document]
Comment (mildaniel) [reply samcli/commands/_utils/cdk_support_decorators.py:34]: Definitely something we can add. I will leave a note about this in the metrics issue.

File: samcli/commands/sync/command.py [hunk 2]
```diff
@@ -142,6 +143,7 @@
 @track_template_warnings([CodeDeployWarning.__name__, CodeDeployConditionWarning.__name__])
 @check_newer_version
 @print_cmdline_args
+@unsupported_command_cdk(alternative_command=None)
```
Comment (jfuss) [inline samcli/commands/sync/command.py:146]: If you make `alternative_command=None` as the default, you don't have to pass in the command unless you need to for specific commands.

File: samcli/commands/package/command.py [hunk 3]
```diff
@@ -81,6 +82,7 @@ def resources_and_properties_help_string():
 @check_newer_version
 @track_template_warnings([CodeDeployWarning.__name__, CodeDeployConditionWarning.__name__])
 @print_cmdline_args
+@unsupported_command_cdk(alternative_command=None)
```
Comment (jfuss) [inline samcli/commands/package/command.py:85]: Should this be running `cdk deploy`? CDK's deploy also does the package.

File: tests/integration/deploy/test_deploy_command.py [hunk 4]
```diff
@@ -1166,6 +1166,39 @@ def test_deploy_update_failed_disable_rollback(self, template_file):
         deploy_process_execute = run_command(deploy_command_list)
         self.assertEqual(deploy_process_execute.process.returncode, 0)
 
+    def test_deploy_logs_warning_with_cdk_project(self):
```
Comment (jfuss) [inline tests/integration/deploy/test_deploy_command.py:1169]: What about package integ tests to validate the correct output is happening?

**Review by jfuss — CHANGES_REQUESTED** [review]

File: samcli/commands/package/command.py (hunk 3 above)
Comment (mildaniel) [reply samcli/commands/package/command.py:85]: I wasn't 100% sure on this, but if you think it's the correct alternative then I'll add that in.

**Review by jfuss — APPROVED** [review]

**Pull request merged**